politeiawww/wsdcrdata.go

```diff
@@ -123,8 +193,88 @@ func newWSDcrdata() (*wsDcrdata, error) {
 	log.Infof("Dcrdata pubsub server version: %v, client version %v",
 		serverSemVer, clientSemVer)
 
+	return c, nil
+}
+
+// reconnect creates a new websocket client, and subscribes to the
+// same subscriptions as the previous client.
+func (w *wsDcrdata) reconnect() error {
+	if w.isShutdown() {
+		return errShutdown
+	}
+
+	prevSubscriptions := make(map[string]struct{}, len(w.subscriptions))
+	timeToWait := 1 * time.Minute
+
+	// Remove existing subscriptions since the client is
+	// no longer connected. These will be resubscribed to.
+	w.client.Stop()
+	for sub := range w.subscriptions {
+		prevSubscriptions[sub] = struct{}{}
+	}
+	w.removeAllSubs()
+
+	for len(prevSubscriptions) > 0 {
+		// Reconnect to dcrdata if needed
+		err := w.ping()
+		if err == errShutdown {
+			return errShutdown
+		} else if err != nil {
+			log.Errorf("wsDcrdata failed to ping dcrdata: %v", err)
+			w.client.Stop()
+
+			// Existing subscriptions have gone bad
+			for sub := range w.subscriptions {
+				prevSubscriptions[sub] = struct{}{}
+			}
+			w.removeAllSubs()
+
+			// Reconnect
+			c, err := newDcrdataWSClient()
+			if err != nil {
+				log.Errorf("wsDcrdata failed to create new client while "+
+					"reconnecting: %v", err)
+				goto wait
+			}
+			w.client = c
+		}
+
+		// Resubscribe
+		for sub := range prevSubscriptions {
+			if w.isSubscribed(sub) {
+				delete(prevSubscriptions, sub)
+				continue
+			}
+			_, err := w.client.Subscribe(sub)
+			if err != nil {
+				log.Errorf("wsDcrdata failed to subscribe to %v: %v",
+					sub, err)
+				goto wait
+			}
+			w.addSub(sub)
+			delete(prevSubscriptions, sub)
+		}
+
+	wait:
+		if len(prevSubscriptions) > 0 {
+			log.Debugf("wsDcrdata reconnect waiting %v", timeToWait)
+			time.Sleep(timeToWait)
+			timeToWait = 2 * timeToWait
```

Comment:
Can we max out an an hour so it doesn't keep backing off forever?

```
			// Increase wait time until it reaches an hour then
			// try to reconnect once per hour.
			timeToWait = 2 * timeToWait
			if timeToWait > 60 {
				timeToWait = 60
			}
```

Reply:
You have to do time.Hour. This will make it 60 nanoseconds.